AWSAppSyncClient/AWSOfflineMutation.swift

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// You may not use this file except in compliance with the License.
+// A copy of the License is located at
+//
+// http://aws.amazon.com/apache2.0
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+
+import Foundation
+
+public enum MutationRecordState: String {
+    case inProgress
+    case inQueue
+    case isDone
+}
+
+public enum MutationType: String {
+    case graphQLMutation
+    case graphQLMutationWithS3Object
+}
+
+final class InternalS3ObjectDetails: AWSS3InputObjectProtocol, AWSS3ObjectProtocol {
+
+    let bucket: String
+    let key: String
+    let region: String
+    let mimeType: String
+    let localUri: String
+
+    init(bucket: String, key: String, region: String, contentType: String, localUri: String) {
+        self.bucket = bucket
+        self.key = key
+        self.region = region
+        self.mimeType = contentType
+        self.localUri = localUri
+    }
+
+    // MARK: AWSS3InputObjectProtocol
+
+    func getLocalSourceFileURL() -> URL? {
+        return URL(string: localUri)
+    }
+
+    func getMimeType() -> String {
+        return mimeType
+    }
+
+    // MARK: AWSS3ObjectProtocol
+
+    func getBucketName() -> String {
+        return bucket
+    }
+
+    func getKeyName() -> String {
+        return key
+    }
+
+    func getRegion() -> String {
+        return region
+    }
+}
+
+final class AWSAppSyncOfflineMutation {
```

Comment:
I think about naming it as AnyMutation. This is because I think that it would be convenient updating offline mutation delegate with new event that just notify about mutation executed. What do you think?

Reply:
But this may require changing properties of the object.

Reply:
Let’s leave the name as it wa for now. But I would like to change how offline mutation delegate works. Now it’s not useful since identifier is generated inside the AppSync, so that API users cannot rely on it. So they even don’t know what identifier is assigned to the mutation.

Reply:
BTw I like the idea how network offline store implemented here https://github.com/Mindera/Alicerce/blob/master/Sources/Stores/NetworkPersistableStore.swift

Reply:
Regarding `AnyMutation`, let's avoid `Any*` patterns except as type-erasing wrappers. For now, I think leaving the name is reasonable, since it's an internal class and I'll probably go back & forth around this code for a couple of iterations.

Reply:
The idea is because `AWSAppSyncOfflineMutationDelegate` is not useful now. You cannot identify Mutation associated outside.
```swift
public protocol AWSAppSyncOfflineMutationDelegate {
    func mutationCallback(recordIdentifier: String, operationString: String, snapshot: Snapshot?, error: Error?)
}
```

So I think that there should be some protocol, like `MutationPersistable` that will allow developers to specify mutation identifier, mutation type, etc. to identify mutation executed. `mutationCallback` will just notify with `AnyMutation` that also conforms `MutationPersistable `.